rfcs/008-map-indexes.md

```diff
@@ -62,15 +62,17 @@ The data model for a map indexed is:
 
 ```
 (<database>, ?DB_VIEWS, <view_signature>, ?VIEW_UPDATE_SEQ) = Sequence
-(<database>, ?DB_VIEWS, <view_signature>, ?VIEW_ID_RANGE, <_id>, <view_id>) = [emitted keys]
+{<database>, ?DB_VIEWS, <view_signature>, ?VIEW_ID_INFO, view_id, ?VIEW_ROW_COUNT} = <row_count>
+{<database>, ?DB_VIEWS, <view_signature>, ?VIEW_ID_INFO, view_id, ?VIEW_KV_SIZE} = <kv_size>
+(<database>, ?DB_VIEWS, <view_signature>, ?VIEW_ID_RANGE, <_id>, <view_id>) = [total_keys, total_size, unique_keys]
 (<database>, ?DB_VIEWS, <view_signature>, ?VIEW_MAP_RANGE, <view_id>, <key>, <_id>, ?ROW_KEYS <count>) = <emitted_keys>
-(<database>, ?DB_VIEWS, <view_signature>, ?VIEW_MAP_RANGE, <view_id>, <key>, <_id>, ?ROW_VALUE, <count>) = <emitted_value>
+(<database>, ?DB_VIEWS, <view_signature>, ?VIEW_MAP_RANGE, <view_id>, {<key>, <_id>}, <count>, ?ROW_VALUE) = <emitted_value>
```

Comment:
I'd also like to understand the reasoning for using two rows for every emitted KV pair. It's a lot of overhead, particularly with the current FDB storage engine which doesn't have prefix elision. Is it simply aiming to preserve the full 100KB value space for user-emitted values?

Reply:
I think this is because of:

> When the emitted keys are encoded to binary to create the FDB key for a row, strings are converted to sort strings and all numbers are converted to doubles. It is not possible to convert either of those encoded values back to the original value.

However, I'm inferring; the spec should be clearer.

Reply:
The sort key transformation is the reason we need to store `<emitted_key>` at all, but I'm conending that a single FDB Value could store both the `<emitted_key>` and `<emitted_value>`.

Reply:
I'm happy to move to a single row that contains `<emitted_key> and <emitted_value>` as long as we ok with the new limits:

* Emitted key must be smaller than 10Kb
* Emitted value must be smaller than 90Kb.

Reply:
@kocolosk The general logic was that previously there was no size limit for key/value sizes so moving to fdb we'd want to preserve as much space as possible to minimize the number of confusing situations where a view on fdb would appear to work but then have a subtle difference on some rows just not appearing without user feedback that there was something amiss.

Originally I hadn't really considered adding size limits to the key/value independently. I was also worried about how to explain to users "The size of the key and value must be less than 100Kib." which was another reason I thought two rows would be better since the documentation would be more straightforward.

Reply:
OK. I don't think that's worth doubling the overhead.

I'd like to leave a buffer so that we've got options for things like storing Values in an encrypted form. We also need a buffer on the key for several reasons:

- sort key generation
- database name (are we really using the raw database name here?)
- view signature

Agreed that it would be bad if views silently crash and omit rows -- at the risk of scope creep do we need to include in the data model the list of documents that failed to index that we had discussed in the past?

Reply:
I'm not against enforcing those limits by any means but we'll want to go through and figure out all of the extra requirements/behaviors and how we communicate those issues to users.

As far as I recall, the tracking documents that failed to be indexed was agreed to in theory but then got hand waved to future version land. I've not got much of an opinion on whether or not its part of this RFC or a later extension. Though I may lean towards later since it's going to have to include user visible API decisions and so on.

Reply:
Hmm, frankly, at this point we're getting to something like:

- 5KB key
- 50KB value

This leaves a decent chunk of overhead for cryptography. We would need to enforce them in the indexing code itself rather than leaving it to FDB (I assume we are doing that anyway, but this pushes us to really need them).